// envio@2.11.6/src/Utils.bs.js
// Generated by ReScript, PLEASE EDIT WITH CARE
'use strict';

var $$Array = require("rescript/lib/js/array.js");
var Js_exn = require("rescript/lib/js/js_exn.js");
var Caml_obj = require("rescript/lib/js/caml_obj.js");
var Belt_Array = require("rescript/lib/js/belt_Array.js");
var Caml_array = require("rescript/lib/js/caml_array.js");
var Belt_Option = require("rescript/lib/js/belt_Option.js");
var Caml_option = require("rescript/lib/js/caml_option.js");
var S$RescriptSchema = require("rescript-schema/src/S.bs.js");

function delay(milliseconds) {
  return new Promise((function (resolve, param) {
                setTimeout((function () {
                        resolve();
                      }), milliseconds);
              }));
}

function mapNone(opt, val) {
  if (opt !== undefined) {
    return ;
  } else {
    return Caml_option.some(val);
  }
}

function catchToNone(unsafeFunc) {
  try {
    return Caml_option.some(unsafeFunc());
  }
  catch (exn){
    return ;
  }
}

function flatten(opt) {
  if (opt !== undefined) {
    return Caml_option.valFromOption(opt);
  }
  
}

function getExn(opt, message) {
  if (opt !== undefined) {
    return Caml_option.valFromOption(opt);
  } else {
    return Js_exn.raiseError(message);
  }
}

var $$Option = {
  mapNone: mapNone,
  catchToNone: catchToNone,
  flatten: flatten,
  getExn: getExn
};

function get(tuple, index) {
  return (tuple[index]);
}

var Tuple = {
  get: get
};

var merge = ((dictA, dictB) => ({...dictA, ...dictB}));

var updateImmutable = ((dict, key, value) => ({...dict, [key]: value}));

var Dict = {
  merge: merge,
  updateImmutable: updateImmutable
};

function minOptInt(a, b) {
  if (a === undefined) {
    if (b !== undefined) {
      return Caml_option.some(Caml_option.valFromOption(b));
    } else {
      return ;
    }
  }
  var a$1 = Caml_option.valFromOption(a);
  if (b !== undefined) {
    return Caml_option.some(Caml_obj.min(a$1, Caml_option.valFromOption(b)));
  } else {
    return Caml_option.some(a$1);
  }
}

var $$Math = {
  minOptInt: minOptInt
};

function mergeSorted(f, xs, ys) {
  if (xs.length === 0) {
    return ys;
  }
  if (ys.length === 0) {
    return xs;
  }
  var n = xs.length + ys.length | 0;
  var result = Array(n);
  var loop = function (_i, _j, _k) {
    while(true) {
      var k = _k;
      var j = _j;
      var i = _i;
      if (i < xs.length && j < ys.length) {
        if (f(Caml_array.get(xs, i), Caml_array.get(ys, j))) {
          Caml_array.set(result, k, Caml_array.get(xs, i));
          _k = k + 1 | 0;
          _i = i + 1 | 0;
          continue ;
        }
        Caml_array.set(result, k, Caml_array.get(ys, j));
        _k = k + 1 | 0;
        _j = j + 1 | 0;
        continue ;
      }
      if (i < xs.length) {
        Caml_array.set(result, k, Caml_array.get(xs, i));
        _k = k + 1 | 0;
        _i = i + 1 | 0;
        continue ;
      }
      if (j >= ys.length) {
        return ;
      }
      Caml_array.set(result, k, Caml_array.get(ys, j));
      _k = k + 1 | 0;
      _j = j + 1 | 0;
      continue ;
    };
  };
  loop(0, 0, 0);
  return result;
}

function setIndexImmutable(arr, index, value) {
  var shallowCopy = arr.slice(0);
  shallowCopy[index] = value;
  return shallowCopy;
}

function transposeResults(results) {
  var _index = 0;
  var output = new Array(results.length);
  while(true) {
    var index = _index;
    if (index >= results.length) {
      return {
              TAG: "Ok",
              _0: output
            };
    }
    var value = results[index];
    if (value.TAG !== "Ok") {
      return value;
    }
    Caml_array.set(output, index, value._0);
    _index = index + 1 | 0;
    continue ;
  };
}

function includes(arr, val) {
  return Belt_Option.isSome(Caml_option.undefined_to_opt(arr.find(function (item) {
                      return Caml_obj.equal(item, val);
                    })));
}

function isEmpty(arr) {
  return arr.length === 0;
}

async function awaitEach(arr, fn) {
  for(var i = 0 ,i_finish = arr.length; i < i_finish; ++i){
    var item = Caml_array.get(arr, i);
    await fn(item);
  }
}

function removeAtIndex(array, index) {
  if (index < 0 || index >= array.length) {
    return $$Array.copy(array);
  }
  var head = array.slice(0, index);
  var tail = Belt_Array.sliceToEnd(array, index + 1 | 0);
  return Belt_Array.concat(head, tail);
}

function last(arr) {
  return Belt_Array.get(arr, arr.length - 1 | 0);
}

function findReverseWithIndex(arr, fn) {
  var _index = arr.length - 1 | 0;
  while(true) {
    var index = _index;
    if (index < 0) {
      return ;
    }
    var item = Caml_array.get(arr, index);
    if (fn(item)) {
      return [
              item,
              index
            ];
    }
    _index = index - 1 | 0;
    continue ;
  };
}

function interleave(arr, separator) {
  var interleaved = [];
  arr.forEach(function (v, i) {
        interleaved.push(v);
        if (i < (arr.length - 1 | 0)) {
          interleaved.push(separator);
          return ;
        }
        
      });
  return interleaved;
}

var $$Array$1 = {
  mergeSorted: mergeSorted,
  setIndexImmutable: setIndexImmutable,
  transposeResults: transposeResults,
  includes: includes,
  isEmpty: isEmpty,
  awaitEach: awaitEach,
  removeAtIndex: removeAtIndex,
  last: last,
  findReverseWithIndex: findReverseWithIndex,
  interleave: interleave
};

function capitalize(str) {
  return str.slice(0, 1).toUpperCase() + str.slice(1);
}

var $$String = {
  capitalize: capitalize
};

function unwrapResultExn(res) {
  if (res.TAG === "Ok") {
    return res._0;
  }
  throw res._0;
}

function $$enum(items) {
  return S$RescriptSchema.union(Belt_Array.mapU(items, S$RescriptSchema.literal));
}

function removeTypeValidationInPlace(schema) {
  schema.f = undefined;
}

function getNonOptionalFieldNames(schema) {
  var acc = [];
  var match = S$RescriptSchema.classify(schema);
  if (typeof match === "object" && match.TAG === "Object") {
    match.items.forEach(function (item) {
          var match = S$RescriptSchema.classify(item.t);
          if (typeof match !== "object") {
            acc.push(item.l);
            return ;
          }
          switch (match.TAG) {
            case "Option" :
            case "Null" :
                return ;
            default:
              acc.push(item.l);
              return ;
          }
        });
  }
  return acc;
}

function getCapitalizedFieldNames(schema) {
  var match = S$RescriptSchema.classify(schema);
  if (typeof match !== "object") {
    return [];
  } else if (match.TAG === "Object") {
    return match.items.map(function (item) {
                return capitalize(item.l);
              });
  } else {
    return [];
  }
}

function coerceToJsonPgType(schema) {
  return S$RescriptSchema.preprocess(schema, (function (s) {
                var match = S$RescriptSchema.classify(s.schema);
                var exit = 0;
                if (typeof match !== "object") {
                  switch (match) {
                    case "Unknown" :
                        exit = 1;
                        break;
                    case "Bool" :
                        exit = 2;
                        break;
                    default:
                      return {};
                  }
                } else {
                  switch (match.TAG) {
                    case "Literal" :
                        if (match._0.kind !== "Null") {
                          return {};
                        }
                        exit = 1;
                        break;
                    case "Null" :
                        exit = 2;
                        break;
                    default:
                      return {};
                  }
                }
                switch (exit) {
                  case 1 :
                      return {
                              s: (function (param) {
                                  return "null";
                                })
                            };
                  case 2 :
                      return {
                              s: (function (unknown) {
                                  if (unknown === null) {
                                    return "null";
                                  } else if (unknown === false) {
                                    return "false";
                                  } else if (unknown === true) {
                                    return "true";
                                  } else {
                                    return unknown;
                                  }
                                })
                            };
                  
                }
              }));
}

var Schema = {
  $$enum: $$enum,
  removeTypeValidationInPlace: removeTypeValidationInPlace,
  getNonOptionalFieldNames: getNonOptionalFieldNames,
  getCapitalizedFieldNames: getCapitalizedFieldNames,
  coerceToJsonPgType: coerceToJsonPgType
};

var $$Set = {};

var $$WeakMap = {};

var $$Map = {};

exports.delay = delay;
exports.$$Option = $$Option;
exports.Tuple = Tuple;
exports.Dict = Dict;
exports.$$Math = $$Math;
exports.$$Array = $$Array$1;
exports.$$String = $$String;
exports.unwrapResultExn = unwrapResultExn;
exports.Schema = Schema;
exports.$$Set = $$Set;
exports.$$WeakMap = $$WeakMap;
exports.$$Map = $$Map;
/* S-RescriptSchema Not a pure module */
